src/probability/martingale/optional_sampling.lean

```diff
@@ -0,0 +1,266 @@
+martingale `f` at `σ` is the conditional expectation of its value at `τ` with respect to the
+σ-algebra genrated by `σ`. -/
+lemma stopped_value_ae_eq_condexp_of_le_of_countable_range
+  [(filter.at_top : filter ι).is_countably_generated]
+  (h : martingale f ℱ μ) (hτ : is_stopping_time ℱ τ) (hσ : is_stopping_time ℱ σ)
+  (hσ_le_τ : σ ≤ τ) (hτ_le : ∀ x, τ x ≤ n)
+  (hτ_countable_range : (set.range τ).countable) (hσ_countable_range : (set.range σ).countable)
+  [sigma_finite (μ.trim hσ.measurable_space_le)] :
+  stopped_value f σ =ᵐ[μ] μ[stopped_value f τ | hσ.measurable_space] :=
+begin
+  haveI : sigma_finite (μ.trim hτ.measurable_space_le),
+  { exact sigma_finite_trim_mono _ (is_stopping_time.measurable_space_mono hσ hτ hσ_le_τ), },
+  have : μ[stopped_value f τ|hσ.measurable_space]
+      =ᵐ[μ] μ[μ[f n|hτ.measurable_space] | hσ.measurable_space],
+    from condexp_congr_ae (h.stopped_value_ae_eq_condexp_of_le_const_of_countable_range hτ hτ_le
+      hτ_countable_range),
+  refine (filter.eventually_eq.trans _ (condexp_condexp_of_le _ hτ.measurable_space_le).symm).trans
+    this.symm,
+  { exact h.stopped_value_ae_eq_condexp_of_le_const_of_countable_range hσ
+      (λ x, (hσ_le_τ x).trans (hτ_le x)) hσ_countable_range, },
+  { exact hσ.measurable_space_mono hτ hσ_le_τ, },
+end
+
+/-- If `τ` and `σ` are two stopping times with `σ ≤ τ` and `τ` is bounded, then the value of a
+martingale `f` at `σ` is the conditional expectation of its value at `τ` with respect to the
+σ-algebra genrated by `σ`. -/
+lemma stopped_value_ae_eq_condexp_of_le [countable ι]
+  (h : martingale f ℱ μ) (hτ : is_stopping_time ℱ τ) (hσ : is_stopping_time ℱ σ)
+  (hσ_le_τ : σ ≤ τ) (hτ_le : ∀ x, τ x ≤ n) [sigma_finite (μ.trim hσ.measurable_space_le)] :
+  stopped_value f σ =ᵐ[μ] μ[stopped_value f τ | hσ.measurable_space] :=
+begin
+  haveI : sigma_finite (μ.trim hτ.measurable_space_le),
+  { exact sigma_finite_trim_mono _ (is_stopping_time.measurable_space_mono hσ hτ hσ_le_τ), },
+  exact h.stopped_value_ae_eq_condexp_of_le_of_countable_range hτ hσ hσ_le_τ hτ_le
+    (set.to_countable _) (set.to_countable _),
+end
+
+end first_countable_topology
+
+section subset_of_nat
+
+/-! In the following results the index set verifies `[linear_order ι] [locally_finite_order ι]` and
+`[order_bot ι]`, which means that it is order-isomorphic to a subset of `ℕ`. `ι` is equipped with
+the discrete topology, which is also the order topology, and is a measurable space with the Borel
+σ-algebra. -/
+
+variables {ι : Type*} [linear_order ι] [locally_finite_order ι] [order_bot ι]
+  [topological_space ι] [discrete_topology ι] [measurable_space ι] [borel_space ι]
+  [measurable_space E] [borel_space E] [second_countable_topology E]
+  {ℱ : filtration ι m} {τ σ : Ω → ι} {f : ι → Ω → E} {i n : ι}
+
+lemma condexp_stopped_value_stopping_time_ae_eq_restrict_le
+  (h : martingale f ℱ μ) (hf_prog : prog_measurable ℱ f)
```

Comment:
Why would you need `prog_measurable` here? You have a martingale in a discrete order setting, so isn't it automatically `prog_measurable`? (`h.adapted.prog_measurable_of_discrete`)